weave/weave_types.py

```diff
@@ -784,7 +785,7 @@ def type_of_instance(cls, obj):
         if not obj:
             return cls(UnknownType())
         list_obj_type = TypeRegistry.type_of(obj[0])
-        for item in obj[1:]:
+        for item in util.sample_rows(obj[1:], 99):
```

Comment:
Fix 4.

I really need confirmation that this is ok to do. As it's currently implemented, this sampling improves performance while reducing reliability for type checking all lists everywhere.